### PR TITLE
Fixes to make life easier with the nlp library

### PR DESCRIPTION
**Pull request by sgugger**
This PR adds two things to make the interface easier with the `nlp` library:
- `BatchEncoding` stops enforcing a 2-dim for every tensor, which causes problems for labels (which should be one vector of shape `[batch_size]`).
- `PreTrainedTokenizerBase.pad` accepts tensors as inputs, which makes it easy to use this function for data collation.

Added proper documentation and tests from @thomwolf initial work.

### CODE REVIEW

File: src/transformers/pipelines.py [hunk 1]
```diff
@@ -2318,7 +2318,7 @@ def _concat_inputs_history(self, inputs: List[List[int]], histories: List[Option
         max_len = max([len(item) for item in outputs])
         outputs = [output + [self.pad_token_id] * (max_len - len(output)) for output in outputs]
         outputs = BatchEncoding(
-            {"input_ids": outputs, "attention_mask": [1] * len(outputs)}, tensor_type=self.framework
+            {"input_ids": outputs, "attention_mask": [[1] * len(outputs)]}, tensor_type=self.framework,
```
Comment (sgugger) [inline src/transformers/pipelines.py:2321]: This is the only thing were the change of dim in BatchEncoding.convert_to_tensors` is breaking something, but in this case, it was a bit magical that the dimension was automatically added, so I don't think this is a serious failure.

**Review by LysandreJik — APPROVED** [review]
LGTM, great that you added tests for all three frameworks.

**Comment by sgugger** [comment]
Merging then, we can follow up next week when @thomwolf is back if he has more comments.
